File: examples/python/echo_bot.py
#!/usr/bin/env python3
"""
WeChat Echo Bot — 完整示例

功能：
  1. 首次运行扫码登录（终端渲染二维码），凭证自动保存
  2. 后续运行自动加载已保存凭证，跳过扫码
  3. 收到消息后显示"正在输入"，然后回复 "Echo: ..."
  4. Session 过期时自动重新扫码

用法：
  uv run examples/python/echo_bot.py
  uv run examples/python/echo_bot.py --force-login

依赖（example 自带，不影响 SDK）：
  qrcode — 终端二维码渲染
"""
from __future__ import annotations

import logging
import sys
import time

from weixin_bot import WeixinBot

# ── 二维码渲染：拦截 SDK 输出的 URL ──────────────────────────────────────

_orig_stderr_write = sys.stderr.write

def _stderr_hook(s: str) -> int:
    if s.startswith("https://") and "qrcode=" in s:
        url = s.strip()
        try:
            import qrcode
            qr = qrcode.QRCode(border=2)
            qr.add_data(url)
            qr.make(fit=True)
            matrix = qr.get_matrix()
            rows = len(matrix)
            for y in range(0, rows, 2):
                line = []
                for x in range(len(matrix[0])):
                    top = matrix[y][x]
                    bot = matrix[y + 1][x] if y + 1 < rows else False
                    if top and bot:
                        line.append("█")
                    elif top and not bot:
                        line.append("▀")
                    elif not top and bot:
                        line.append("▄")
                    else:
                        line.append(" ")
                _orig_stderr_write("".join(line) + "\n")
        except ImportError:
            pass
    return _orig_stderr_write(s)

sys.stderr.write = _stderr_hook  # type: ignore[assignment]

# ── 日志 ──────────────────────────────────────────────────────────────────

logging.basicConfig(level=logging.INFO, format="%(message)s")

def log(level: str, msg: str) -> None:
    ts = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime())
    logging.info("%s [%s] %s", ts, level, msg)

# ── 启动 ──────────────────────────────────────────────────────────────────

force_login = "--force-login" in sys.argv

bot = WeixinBot(on_error=lambda err: log("ERROR", str(err)))

log("INFO", "强制重新扫码登录..." if force_login else "正在登录（已有凭证则自动跳过扫码）...")
creds = bot.login(force=force_login)
log("INFO", f"登录成功 — Bot ID: {creds.account_id}")
log("INFO", f"关联用户: {creds.user_id}")
log("INFO", f"API 地址: {creds.base_url}")

message_count = 0
start_time = time.time()


@bot.on_message
async def handle(msg):
    global message_count
    message_count += 1
    elapsed = int(time.time() - start_time)

    log("RECV", f"#{message_count} | 类型: {msg.type} | 用户: {msg.user_id}")
    log("RECV", f"内容: {msg.text}")

    try:
        await bot.send_typing(msg.user_id)
    except Exception:
        pass

    reply = f"Echo: {msg.text}"

    try:
        await bot.reply(msg, reply)
        log("SEND", f"回复成功 ({len(reply)} 字符) | 运行 {elapsed}s | 累计 {message_count} 条")
    except Exception as err:
        log("ERROR", f"回复失败: {err}")


log("INFO", "开始接收微信消息 (Ctrl+C 停止)")
log("INFO", "────────────────────────────────────")

try:
    bot.run()
except KeyboardInterrupt:
    log("INFO", f"Bot 已停止，共处理 {message_count} 条消息")

File: examples/nodejs/echo-bot.ts
#!/usr/bin/env npx tsx
/**
 * WeChat Echo Bot — 完整示例
 *
 * 功能：
 *   1. 首次运行扫码登录，凭证自动保存到 ~/.weixin-bot/credentials.json
 *   2. 后续运行自动加载已保存凭证，跳过扫码
 *   3. 收到微信消息后原样回复 "Echo: ..."
 *   4. Session 过期时自动重新扫码
 *
 * 用法：
 *   npx tsx examples/nodejs/echo-bot.ts
 *   npx tsx examples/nodejs/echo-bot.ts --force-login   # 强制重新扫码
 */

import { WeixinBot } from '@pinixai/weixin-bot'

// ── 配置 ─────────────────────────────────────────────────────────────────

const forceLogin = process.argv.includes('--force-login')

// ── 日志工具 ──────────────────────────────────────────────────────────────

function log(level: string, msg: string) {
  const ts = new Date().toISOString()
  console.log(`${ts} [${level}] ${msg}`)
}

// ── 启动 ──────────────────────────────────────────────────────────────────

const bot = new WeixinBot({
  onError: (err) => {
    log('ERROR', err instanceof Error ? err.stack ?? err.message : String(err))
  },
})

// 登录
log('INFO', forceLogin ? '强制重新扫码登录...' : '正在登录（已有凭证则自动跳过扫码）...')
const creds = await bot.login({ force: forceLogin })
log('INFO', `登录成功 — Bot ID: ${creds.accountId}`)
log('INFO', `关联用户: ${creds.userId}`)
log('INFO', `API 地址: ${creds.baseUrl}`)

// 统计
let messageCount = 0
const startTime = Date.now()

// 注册消息处理
bot.onMessage(async (msg) => {
  messageCount++
  const elapsed = ((Date.now() - startTime) / 1000).toFixed(0)

  log('RECV', `#${messageCount} | 类型: ${msg.type} | 用户: ${msg.userId}`)
  log('RECV', `内容: ${msg.text}`)
  log('RECV', `context_token: ${msg._contextToken.slice(0, 20)}...`)

  const reply = `Echo: ${msg.text}`

  try {
    await bot.reply(msg, reply)
    log('SEND', `回复成功 (${reply.length} 字符) | 运行 ${elapsed}s | 累计 ${messageCount} 条`)
  } catch (err) {
    log('ERROR', `回复失败: ${err instanceof Error ? err.message : String(err)}`)
  }
})

// 优雅退出
process.on('SIGINT', () => {
  log('INFO', `收到 SIGINT，正在停止... (共处理 ${messageCount} 条消息)`)
  bot.stop()
})

process.on('SIGTERM', () => {
  log('INFO', `收到 SIGTERM，正在停止...`)
  bot.stop()
})

// 启动长轮询
log('INFO', '开始接收微信消息 (Ctrl+C 停止)')
log('INFO', '────────────────────────────────────')
await bot.run()

log('INFO', `Bot 已停止，共处理 ${messageCount} 条消息`)
